File: src/main/player.ts
import { BrowserWindow, ipcMain } from 'electron'
import { loadConfig, updateConfig } from './config'
import { AudioDevice, PlaybackState, PlayerCommand } from '@shared/types'

let win: BrowserWindow | null = null
let devices: AudioDevice[] = []
const state: PlaybackState = {
  trackId: null,
  playing: false,
  position: 0,
  duration: 0,
  volume: 1
}

// Hooks other modules subscribe to (the queue engine wires these in M5).
let endedHandler: (() => void) | null = null
let stateListener: ((s: PlaybackState) => void) | null = null
let readyHandler: (() => void) | null = null

export function setPlayerWindow(w: BrowserWindow): void {
  win = w
}

export function onTrackEnded(fn: () => void): void {
  endedHandler = fn
}

export function onStateChange(fn: (s: PlaybackState) => void): void {
  stateListener = fn
}

export function onReady(fn: () => void): void {
  readyHandler = fn
}

function send(cmd: PlayerCommand): void {
  win?.webContents.send('player:command', cmd)
}

function emitState(): void {
  stateListener?.({ ...state })
}

/** Registers the IPC listener for events reported by the player renderer. */
export function initPlayerIpc(): void {
  ipcMain.on('player:event', (_e, event: string, payload: unknown) => {
    switch (event) {
      case 'ready': {
        // Renderer (re)loaded: refresh devices and apply the saved output.
        send({ type: 'enumerate' })
        const saved = loadConfig().outputDeviceId
        if (saved) send({ type: 'setSinkId', deviceId: saved })
        readyHandler?.()
        break
      }
      case 'devices':
        devices = (payload as AudioDevice[]) ?? []
        break
      case 'timeupdate': {
        const p = payload as Partial<PlaybackState>
        if (typeof p.position === 'number') state.position = p.position
        if (typeof p.duration === 'number') state.duration = p.duration
        if (typeof p.playing === 'boolean') state.playing = p.playing
        if (typeof p.trackId === 'number') state.trackId = p.trackId
        emitState()
        break
      }
      case 'ended':
        state.playing = false
        emitState()
        endedHandler?.()
        break
      case 'error':
        console.error('[player] renderer:', payload)
        break
    }
  })
}

// ---- Transport (used by the queue engine and admin controls) ----------------

export function loadTrack(trackId: number, autoplay = true): void {
  state.trackId = trackId
  state.position = 0
  state.duration = 0
  state.playing = autoplay
  send({ type: 'load', trackId, autoplay })
  emitState()
}

export function play(): void {
  state.playing = true
  send({ type: 'play' })
  emitState()
}

export function pause(): void {
  state.playing = false
  send({ type: 'pause' })
  emitState()
}

export function seek(position: number): void {
  send({ type: 'seek', position })
}

export function setVolume(value: number): void {
  state.volume = Math.min(Math.max(value, 0), 1)
  send({ type: 'volume', value: state.volume })
  emitState()
}

export function setOutputDevice(deviceId: string): void {
  updateConfig({ outputDeviceId: deviceId || null })
  send({ type: 'setSinkId', deviceId })
}

export function requestDevices(): void {
  send({ type: 'enumerate' })
}

export function getDevices(): AudioDevice[] {
  return devices
}

export function getState(): PlaybackState {
  return { ...state }
}

File: crates/jukebox-server/src/routes/player.rs
//! `/api/player/*`: playback state and the admin's transport controls.

use axum::extract::State;
use axum::http::{HeaderMap, StatusCode};
use axum::response::Response;
use jukebox_core::js;
use jukebox_core::types::{DevicesResponse, OutputResponse};
use serde_json::Value;

use super::{blocking, internal, require_admin, AppStateRef};
use crate::http::{error, ok, JsonBody};

pub async fn state(State(state): AppStateRef) -> Response {
    ok(&state.player.state())
}

pub async fn devices(State(state): AppStateRef, headers: HeaderMap) -> Response {
    if let Err(denied) = require_admin(&state, &headers) {
        return denied;
    }
    state.player.request_devices();
    ok(&DevicesResponse {
        devices: state.player.devices(),
        selected: state.config.get().output_device_id,
    })
}

pub async fn output(State(state): AppStateRef, headers: HeaderMap, body: JsonBody) -> Response {
    if let Err(denied) = require_admin(&state, &headers) {
        return denied;
    }
    let device_id = body.string_or_empty("deviceId");
    let config = state.config.clone();
    let id = device_id.clone();
    let saved =
        blocking(move || config.update(|c| c.output_device_id = (!id.is_empty()).then_some(id)))
            .await;
    if let Err(err) = saved {
        return internal(err);
    }
    state.player.set_output_device(&device_id);
    ok(&OutputResponse {
        ok: true,
        selected: state.config.get().output_device_id,
    })
}

pub async fn load(State(state): AppStateRef, headers: HeaderMap, body: JsonBody) -> Response {
    if let Err(denied) = require_admin(&state, &headers) {
        return denied;
    }
    let track_id = body.number("trackId");
    if !js::is_integer(track_id) {
        return error(StatusCode::BAD_REQUEST, "trackId required");
    }
    let autoplay = body.get("autoplay") != Some(&Value::Bool(false));
    state.player.load(track_id as i64, autoplay);
    ok(&state.player.state())
}

pub async fn play(State(state): AppStateRef, headers: HeaderMap) -> Response {
    if let Err(denied) = require_admin(&state, &headers) {
        return denied;
    }
    state.player.play();
    ok(&state.player.state())
}

pub async fn pause(State(state): AppStateRef, headers: HeaderMap) -> Response {
    if let Err(denied) = require_admin(&state, &headers) {
        return denied;
    }
    state.player.pause();
    ok(&state.player.state())
}

pub async fn seek(State(state): AppStateRef, headers: HeaderMap, body: JsonBody) -> Response {
    if let Err(denied) = require_admin(&state, &headers) {
        return denied;
    }
    let position = body.number("position");
    if !position.is_finite() {
        return error(StatusCode::BAD_REQUEST, "position required");
    }
    state.player.seek(position);
    ok(&state.player.state())
}

pub async fn volume(State(state): AppStateRef, headers: HeaderMap, body: JsonBody) -> Response {
    if let Err(denied) = require_admin(&state, &headers) {
        return denied;
    }
    let value = body.number("value");
    if !value.is_finite() {
        return error(StatusCode::BAD_REQUEST, "value required");
    }
    state.player.set_volume(value);
    ok(&state.player.state())
}
